### ISSUE TITLE
Temporal/market-hour features improve GAN-augmented strategy performance significantly

### ISSUE THREAD

**Issue by viking76** [nateemma/strategies#58]
### Description

The current `BaseNNStrategy` has calendar features (`hour_sin/cos`, `dow_sin/cos`, etc.) commented out with a note citing low correlation with buy/sell labels. However, **richer temporal features encoding market microstructure** combined with a **custom adaptive Fibonacci indicator** can substantially improve performance when used with GAN-augmented NN strategies.

### Proposed features (from `Framework/TemporalFeatures.py`)

8 columns computed in Paris local time with DST handling via `zoneinfo`:

| Feature | Type | Description |
|---|---|---|
| `hour_sin`, `hour_cos` | float | Cyclic hour encoding (passthrough for GAN) |
| `is_weekend` | binary | Weekend flag |
| `is_asia_open` | binary | Asia session (01:00–09:00 Paris) |
| `is_eu_open` | binary | EU session (09:00–17:30) |
| `is_us_open` | binary | US session (14:30–21:00) |
| `is_market_open` | binary | Any session open |
| `market_session` | categorical | `asia`/`eu`/`us`/`none` |

Session indicators set to 0 on weekends regardless of hour.

### Fibodynamique indicator

Port of NinjaTrader's Fibodynamic — an **adaptive Fibonacci retracement** that dynamically adjusts its lookback period based on the average time between swing highs/lows, resetting when price crosses the 50% level. Produces 5 levels (`Fib0`, `Fib382`, `Fib500`, `Fib618`, `Fib1000`) usable as model features alongside the temporal columns.

### Implementation notes

- `add_temporal_features(dataframe)` — vectorized, single-pass, no loops
- Resolves index from DatetimeIndex or `"date"` column, localizes from UTC to `Europe/Paris`
- `hour_sin/cos` go in `gan_passthrough_columns` (same as `dow_sin/cos`) to prevent the GAN from generating invalid cyclic pairs
- `market_session` one-hot encoded to 3 booleans for the model
- Uses **TabDDPM** — a Tabular Denoising Diffusion Probabilistic Model, like DDPM but designed for tabular data and compatible with x86 hardware (CPU/GPU, no Apple MLX dependency)

### Results (NNNC_TabDDPM_LSTM3, 15 USDC pairs, 15m, ~15 months)

Backtest comparison (30 features → 37 after temporal + Fibodynamique, GAN retrained):

| Metric | Before (30 feat) | After (37 feat) | Change |
|---|---|---|---|
| Total profit | ~14.2% | **19.08%** | **+34%** |
| Win rate | 80.5% | 68.4% | More selective, higher quality |
| Sharpe | baseline | **+44%** | |
| GAN best loss | 0.5116 | **0.4121** | **−19%** |

Hyperopt best epoch: **43.85% profit**, 320 trades, 68.4% win rate, max DD 8.89%, loss = −1.615.

### Why these work when raw `hour_sin/cos` didn't

1. **Session boundaries encode market microstructure** — crypto activity clusters around Asia open, EU overlap (high liquidity), US open (volatility spike). Binary session flags give the classifier clear regime switches.
2. **Weekend flag** prevents false signals during low-liquidity periods where technical indicators are unreliable.
3. **DST-aware** — Paris time automatically handles EU/US session boundary shifts, preventing a recurring seasonal bias.
4. **GAN passthrough** prevents TabDDPM from generating invalid temporal patterns (e.g., `is_us_open=1` at 03:00 Paris time).
5. **Fibodynamique** provides adaptive support/resistance levels that capture changing market volatility — more responsive than fixed-period Fibonacci.

### Suggested integration

1. Add `Framework/TemporalFeatures.py` (60 lines, standalone, no new deps beyond stdlib `zoneinfo`)
2. Import and call `add_temporal_features()` in `BaseNNStrategy.add_additional_indicators()` or per-strategy
3. Add `_fibodynamic()` indicator alongside temporal columns
4. Add the 7 new temporal columns + Fib levels to `include_list` and `pre_normalized_columns`; add `hour_sin/cos` to `gan_passthrough_columns`
5. Retrain GAN and classifier with the expanded feature set

### Fibodynamique indicator

Port of NinjaTrader's Fibodynamic — an **adaptive Fibonacci retracement** that works like standard Fib levels but with two key innovations:

1. **Adaptive lookback** — instead of a fixed period, the swing high/low detection window adjusts dynamically based on the average bar count between recent swing points. In fast markets the lookback shortens, in slow markets it lengthens.
2. **50%-level reset** — when price crosses the 0.50 Fib level, the reference swing high/low reset immediately. This prevents stale levels from persisting after a trend reversal, keeping the indicator responsive.

Produces 5 levels per bar (`Fib0` / `Fib382` / `Fib500` / `Fib618` / `Fib1000`) usable directly as model features alongside the temporal columns.
Happy to submit a PR if there's interest.

**Comment by nateemma** [comment]
Hi,

  so, I tried this out and didn't find any improvement at all. My suspicion
is that your fibonacci code is leaking future data (e.g. by using a
centered window). Could you either send me your code for that, or run this
command:

bash user_data/strategies/scripts/check_bias.sh -n 30 NNNC
NNNC_TabDDPM_LSTM3

Thanks,

Phil

**Comment by viking76** [comment]
I work with Ubuntu in x86 / 3070ti so i have boosted my GAN to make synthesis data. Sorry, sometime the code is in french. Somes dep files I hade indicator of open us eur market, to indicate the perception of volatility in this time. i trade on binance international with USDC base 
[NNNC_TabDDPM_LSTM3_deps.zip](https://github.com/user-attachments/files/29522254/NNNC_TabDDPM_LSTM3_deps.zip)

**Comment by nateemma** [comment]
Hi Gael,

  I tried your changes and did not find any improvement at all - in fact
the temporal indicators made performance worse. I tried both MLX and
tensorflow versions, and different model types, with and without GAN.

Some questions:

- are your numbers from hyperopt or backtesting?
- if backtesting, did you test a timeframe inside or outside the hyperopt
timeframe? If inside, can you try a timeframe outside the hyperopt period?
- if you have backtest data, can you look at the monthly gains? You may
have gains concentrated in a few months (November 2025 usually causes
issues), which would indicate that the model is overfitted for that period.

Thanks,

Phil

**Comment by viking76** [comment]
Hi Phil,

I have tested with backtesting with long time range inside hyperopt and maybe overfitted i think.
Sorry, to have losted your time.
 thks Gaël

**Comment by nateemma** [comment]
not at all, it seemed feasible

hanks,

Phil

**Comment by nateemma** [comment]
I'll take a look, but it will be a few weeks before I can do that because
I'm in the middle of a different project - check out repo
fx-strategy-framework if you are interested.

Also, after your last email, I ran a lot of statistical analysis on various
features (temporal and otherwise). Results are in us_spot_market_study.md -
but the caveat is that this is for the US market (binanceus), likely a
different result for global binance. The conclusion of that is that all of
the gains are delivered by 1 coin (ZEC), even then only in certain bullish
regimes, and that it is statistically impossible to beat the market using
only the candle data (for prediction algorithms anyway, without access to
short trades). Note that the NNNC/NNMT strategies are still profitable,
just not competitive against a simple buy and hold strategy.

As an aside, this led me to look at various basket strategies (in directory
Basket), which are actually fairly profitable. One in particular
(MomentumRegimeBasket15m) looks quite promising, but operates on a
timeframe of months.

Anyway, when I have time, I'll create a model variant based on your
coconut suggestion and test it against the default indicators and your
additions

Thanks,

Phil

**Comment by viking76** [comment]
I found again a data leakage in coconut model; I'm fixing it and will set up a repository on my end. I went through the code for your other ongoing project. If it's intended for Forex? you don't have any connections to trading platform APIs yet. I used to trade the Dow Jones with prop firms.

**Comment by nateemma** [comment]
FYI, I tried several coconut variations - there's a report at
docs/superpowers/specs/2026-07-18-noisycoconut-nnnc-design.md
(I really should move that). I keep hitting issues with shallow volume for
US markets, but it might work for Binance Global

thanks,

Phil
